doc/adr/0021-undo-redo.md

```diff
@@ -0,0 +1,101 @@
+
+[Operations in Athens](doc/adr/0010-atomic-composite-grapth-operations.md) follow the principle of determinism, where the operation itself carries enough information for all participants to perform the same changes with no ambiguity over the same state. This property enables us to derive an undo operation from a state together with any operation.
+
+We can reduce the participants to three types: the undo issuer, the server, and other users. Although in theory any of these could construct the concrete undo operation, it is onerous for the server and other users to keep all possible past states that might be required for undos from anyone, for any point in time.
+
+The prime candidate to keep relevant states is the undo issuer themselves. This participant can keep a list of states needed for their own set of undoable operations. However, this operation is still optimistic in nature, and subject to conflicts. It is tantamount to saying "this is how to undo the operation given what I know now".
+
+We can also transfer the burden of resolution to the server, achieving a "true" undo since the server itself is the source of truth for the RTC system. The server would still need a way to keep or recall past states in order to resolve the undo operation. This approach would require the client issuing the undo to somehow direct the server to undo the operation, and wait for the result.
+
+
+## Approach
+
+The primary motivation for undo/redo right now is the prevention of data loss in the moment, and keeping that goal in mind we decided to pursue the first candidate (issuer resolves undo operation). Higher fidelity reconstruction of paste application state, either for inspection or data recovery, is left to future time travel functionality.
+
+For each atomic operation over a db state, we can build a corresponding atomic or composite that undoes it. For composite operations, we can compute the undo atomic/composite for each atomic operation in it, maintaining order so that the operations make sense. A redo operation is the undo operation for the effected undo.
+
+This is the list of undo operations for each atomic:
+  - undo operation map
+      - block
+          - :block/new
+              - undo is :block/remove for uid
+          - :block/save
+              - undo is :block/save to previous str
+          - :block/open
+              - undo is :block/open with inverse bool
+          - :block/remove
+              - undo is
+                  - paste of previous IR for block uid in previous position
+                  - :block/save for all the ref in str that were replaced in remove
+          - :block/move
+              - undo is move to previous position
+      - page
+          - :page/new
+              - undo is :page/remove for title
+          - :page/rename
+              - undo is :page/rename to previous title
```

Comment:
This is a rare case, that we guard for not happening, but still worth considering here.
What if initial `:page/rename` renamed to a name that had unlinked `[[new page name]]`?
When we undo, we shouldn't fix previously wrong state.

Reply:
We could do the more sophisticated resolution where we only rename the refs that were changed in the original operation. We have all the data for this, and we do it for merge and remove.

I think we might as well aim to do it here and if we compromise to reduce complexity then we add that to the `Insights from MVP` section.

Reply:
Checked the source for the current rename, and it doesn't look like unlinked refs ever get updated on the rename. So the undo will not change those.